app/models/miq_server/worker_management/monitor/kubernetes.rb

```diff
@@ -0,0 +1,116 @@
+    current_pods.select { |_name, h| h[:last_state_terminated] && h.fetch(:container_restarts, 0) > restart_count }.collect { |_name, h| h[:label_name] }
+  end
+
+  private
+
+  def start_pod_monitor
+    @monitor_thread ||= begin
+      Thread.new { monitor_pods }
+    end
+  end
+
+  def ensure_pod_monitor_started
+    if @monitor_thread.nil? || !@monitor_thread.alive?
+      if !@monitor_thread.nil? && @monitor_thread.status.nil?
+        dead_thread, @monitor_thread = @monitor_thread, nil
+        _log.info("Waiting for the Monitor Thread to exit...")
+        dead_thread.join
+      end
+
+      start_pod_monitor
+    end
+  end
+
+  def delete_failed_deployments
+    # TODO: We should have a list of worker deployments we'll delete to avoid accidentally killing pg/memcached/orchestrator
+    # See ContainerOrchestrator#get_pods
+    failed_deployments.each do |failed|
+      orchestrator.delete_deployment(failed)
+    end
+  end
+
+  def orchestrator
+    @orchestrator ||= ContainerOrchestrator.new
+  end
+
+  def monitor_pods
+    loop do
+      reset_current_pods
+      collect_initial_pods
+
+      # watch_for_pod_events doesn't return unless an error caused us to break out of it, so we'll reset and start over again
+      watch_for_pod_events
+    end
+  end
+
+  def collect_initial_pods
+    pods = orchestrator.get_pods
+    pods.each { |p| save_pod(p) }
+    self.pod_resource_version = pods.resourceVersion
+  end
+
+  def watch_for_pod_events
+    watcher = orchestrator.watch_pods(pod_resource_version) do |event|
+      case event.type.downcase
+      when "added", "modified"
+        save_pod(event.object)
+      when "deleted"
+        delete_pod(event.object)
+      when "error"
+        if status = event.object
+          # ocp 3 appears to return 'ERROR' watch events with the object containing the 410 code and "Gone" reason like below:
+          # #<Kubeclient::Resource type="ERROR", object={:kind=>"Status", :apiVersion=>"v1", :metadata=>{}, :status=>"Failure", :message=>"too old resource version: 199900 (27177196)", :reason=>"Gone", :code=>410}>
+          log_pod_error_event(status.code, status.message, status.reason)
+        end
+
+        self.pod_resource_version = 0
+        break
+      end
+    end
+  ensure
+    watcher.finish if watcher
+  end
+
+  def log_pod_error_event(code, message, reason)
+    _log.warn("Restarting watch_pods at resource_version 0 due to error: [#{code} #{reason}], [#{message}]")
+  end
+
+  def save_pod(pod)
+    return unless pod.status.containerStatuses
+
+    ch = Concurrent::Hash.new
+    ch[:label_name]            = pod.metadata.labels.name
+    ch[:last_state_terminated] = pod.status.containerStatuses.any? { |cs| !!cs.lastState.terminated }
+    ch[:container_restarts]    = pod.status.containerStatuses.inject(0) { |sum, cs| sum += cs.restartCount if cs.lastState.terminated; sum }
+
+    name = pod.metadata.name
+    if current_pods[name]
+      current_pods[name].merge!(ch)
```

Comment:
something about this feels un-thread-safe since 2 threads could could go a getter and then a getter again, and then a merge and they could be all out of order.  I don' think it actually matter because the only writer should be the monitor thread, and the main thread is only a reader (except before the monitor thread, but then we're ok).

There's no method for check-and-set on Concurrent::Hash, but there is on Concurrent::Map, so maybe that's a better choice?

With Concurrent::Hash, perhaps this is cleaner:

```ruby
current_pods[name] ||= ch
current_pods[name].merge!(ch)
```

Reply:
We now only have the monitor thread doing writes to the concurrent hash and we ensure it's dead before starting a new thread so it should have only 1 writer, but x readers.  If we had multiple writers, I'd probably throw them into a queue for a single thread to process and update the `current_pods` in a single thread, let we do in the kubernetes provider catcher.

Reply:
Although, it's needed yet since there's only one writing thread, I like your simple workaround as a way to remain thread-safe from the start.